packages/flutter_tools/lib/src/daemon.dart

```diff
@@ -3,108 +3,192 @@
-  Future<void> dispose() async { }
+  /// The data of the event.
+  final Object? data;
+
+  /// Stream of the binary content of the event.
+  ///
+  /// Must be listened to if binary data is present.
+  final Stream<List<int>>? binary;
 }
 
-/// A [DaemonStream] that uses stdin and stdout as the underlying streams.
-class StdioDaemonStreams extends DaemonStreams {
-  StdioDaemonStreams(Stdio stdio) :
-    _stdio = stdio,
-    inputStream = _convertInputStream(stdio.stdin);
+const String _binaryLengthKey = '_binaryLength';
 
-  final Stdio _stdio;
+enum _InputStreamParseState {
+  JSON,
+  BINARY,
+}
 
-  @override
-  final Stream<Map<String, Object?>> inputStream;
+/// Converts a binary stream to a stream of [DaemonMessage].
+///
+/// The daemon JSON-RPC protocol is defined as follows: every single line of
+/// text that starts with `[{` and ends with `}]` will be parsed as a JSON
+/// message. The array should contain only one single object which contains the
+/// message data.
+///
+/// If the JSON object contains the key [_binaryLengthKey] with an integer
+/// value (will be refered to as N), the following N bytes after the newline
+/// character will contain the binary part of the message.
+@visibleForTesting
+Stream<DaemonMessage> convertInputStream(Stream<List<int>> inputStream) async* {
+  _InputStreamParseState state = _InputStreamParseState.JSON;
 
-  @override
-  void send(Map<String, Object?> message) {
-    _stdio.stdoutWrite(
-      '[${json.encode(message)}]\n',
-      fallback: (String message, Object? error, StackTrace stack) {
-        throwToolExit('Failed to write daemon command response to stdout: $error');
-      },
-    );
-  }
-}
+  const int LF = 10; // The '\n' character
 
-/// A [DaemonStream] that uses [Socket] as the underlying stream.
-class TcpDaemonStreams extends DaemonStreams {
-  /// Creates a [DaemonStreams] with an existing [Socket].
-  TcpDaemonStreams(
-    Socket socket, {
-    required Logger logger,
-  }): _logger = logger {
-    _socket = Future<Socket>.value(_initializeSocket(socket));
+  final BytesBuilder bytesBuilder = BytesBuilder(copy: false);
+  late StreamController<List<int>> currentBinaryStream;
+  int remainingBinaryLength = 0;
+  await for (final List<int> chunk in inputStream) {
+    int start = 0;
+    while (start < chunk.length) {
+      if (state == _InputStreamParseState.JSON) {
+        // Search for newline character.
+        final int indexOfNewLine = chunk.indexOf(LF, start);
+        if (indexOfNewLine < 0) {
+          bytesBuilder.add(chunk.sublist(start));
+          start = chunk.length;
+        } else {
+          bytesBuilder.add(chunk.sublist(start, indexOfNewLine + 1));
+          start = indexOfNewLine + 1;
+
+          // Process chunk here
+          final Uint8List combinedChunk = bytesBuilder.takeBytes();
+          String jsonString = utf8.decode(combinedChunk).trim();
+          if (jsonString.startsWith('[{') && jsonString.endsWith('}]')) {
```

Comment:
Maybe handle error case where message is malformed?

Reply:
I preserved the original behavior for backward compatibility. In the original daemon, anything that is not JSON string are simply ignored.